Add --runtime-arg option for run and serve

edmcman opened the pull request:
See #948 

Useful for passing `--jinja` to llama.cpp at the moment to enable tool calling.

## Summary by Sourcery

Adds the `--runtime-args` option to the `run` and `serve` commands, allowing users to pass arguments directly to the underlying runtime (llama.cpp or vllm).

New Features:
- Adds the `--runtime-args` option to the `run` and `serve` commands.

Tests:
- Adds tests to verify that `--runtime-args` are passed to the runtime.

engelmi commented on ramalama/cli.py:
```diff
@@ -690,6 +690,14 @@ def run_serve_perplexity_args(parser):
+        action='append',
+        type=str,
+        default=[],
+        help="argument to add to runtime invocation"
```
NIT:
```suggestion
        help="arguments to add to runtime invocation"
```

engelmi commented on ramalama/cli.py:
```diff
@@ -690,6 +690,14 @@ def run_serve_perplexity_args(parser):
+        "--runtime-arg",
+        dest="runtime_args",
+        action='append',
+        type=str,
```
```suggestion
        type=str,
        nargs="+",
```

engelmi commented on ramalama/cli.py:
```diff
@@ -690,6 +690,14 @@ def run_serve_perplexity_args(parser):
         help="size of the prompt context (0 = loaded from model)",
     )
+    parser.add_argument(
+        "--runtime-arg",
```
NIT:
```suggestion
        "--runtime-args",
```

engelmi reviewed (commented):
I think some of the end-to-end tests need to be updated regarding the expected output.

rhatdan commented:
Needs man page documention?

Sourcery has some good points.

Test required, basically just look at the dryrun tests.

rhatdan commented:
Thanks @edmcman

rhatdan commented:
You can run `make validate` locally to descover quickly issues that will prevent your PR from being merged.

ericcurtin commented:
You need to sign your commits to pass the DCO build, recommend reading the build logs for the others.

edmcman commented:
Thanks, I will try to address the other issues after work tonight.

edmcman commented:
FYI [ramalama-serve.md](https://github.com/containers/ramalama/blob/main/docs/ramalama-serve.1.md) seems pretty specific to llama.cpp.  Is this right?  Does `ramalama-serve` not support `vllm`?

edmcman commented:
Ah, I don't think `nargs="+"` will work here because we might want to pass things that look like arguments to `argparse`.  The only way I know how to do this is too have `nargs=1` and use `--runtime-args=--arg1 --runtime-args=--arg2` and use append.  But if there is a better way I am all ears...

engelmi commented:
> Ah, I don't think `nargs="+"` will work here because we might want to pass things that look like arguments to `argparse`. The only way I know how to do this is too have `nargs=1` and use `--runtime-args=--arg1 --runtime-args=--arg2` and use append. But if there is a better way I am all ears...

Makes sense. With `nargs="+"` and `action="append"` we'll get a list of lists with one element.

ericcurtin commented:
There are other implementations of arg appending tools, dracut is one off the top of my head (it's kargs, but the same concept), might be worth a little bit of research on existing implementations. Arg parsing, appending, etc. can get hairy (especially when adding multiple args at once), would be nice to get it right the first time.

ericcurtin commented:
> FYI [ramalama-serve.md](https://github.com/containers/ramalama/blob/main/docs/ramalama-serve.1.md) seems pretty specific to llama.cpp. Is this right? Does `ramalama-serve` not support `vllm`?

It can also be used with vllm.

edmcman commented:
> Well we could not support a list, and just force users to put all options into one command, which should work.
> 
> --runtime-arg="--arg1 --arg2=foo --arg3"

We then have to parse the the string to split it.  Though maybe that is easy with [this](https://docs.python.org/3/library/shlex.html#shlex.split).  I kind of like that plan, though as @ericcurtin suggested I'll take a look at dracut.

edmcman commented:
@ericcurtin Were you thinking of `--kernel-cmdline`?  I guess those don't really have dashes in them, so it seems like they didn't have to worry about that problem.

https://github.com/dracutdevs/dracut/blob/5d2bda46f4e75e85445ee4d3bd3f68bf966287b9/test/test-functions#L29

edmcman commented:
Can someone allow the CI to run the BATS test again?  I am having trouble running it locally.

edmcman commented:
Hmm.  I think there might be an issue with ` run_ramalama` passing the quoted argument properly...

edmcman commented:
Woohoo. My BATS tests are finally passing :sweat:  I don't think the current bats failures are my fault.  Seems like a flaky test?

ericcurtin commented:
> @ericcurtin Were you thinking of `--kernel-cmdline`? I guess those don't really have dashes in them, so it seems like they didn't have to worry about that problem.
> 
> https://github.com/dracutdevs/dracut/blob/5d2bda46f4e75e85445ee4d3bd3f68bf966287b9/test/test-functions#L29

I think I was actually thinking of grubby 😅  I think that can add kargs.

It was just the first example of a arg appending tool I could think of. There may be some pitfalls to consider like how we interpret " , etc.

edmcman commented:
Looks like grubby calls `poptParseArgvString` https://github.com/rhboot/grubby/blob/c01b0d5bb182bde35b464d14996acf354a3ada2e/grubby.c#L3804

> Although popt is usually used for parsing arguments already divided into an argv-style array, some programs need to parse strings that are formatted identically to command lines. To facilitate this, popt provides a function that parses a string into an array of strings, using rules similiar to normal shell parsing.

This sounds like `shlex.split`, which I am using now.

My thoughts:
1. Add a test or two involving quoted arguments
2. Move the `shlex.split` calls closer to the argument parsing, rather than in `model.py` where they are currently.

rhatdan commented:
You need to squash your commits.

git rebase -i origin 
git push --force

rhatdan commented:
LGTM
